cdc/cdc.cc

```diff
@@ -747,33 +758,62 @@ transform_mutations(std::vector<mutation>& muts, decltype(muts.size()) batch_siz
-            muts.reserve(2 * muts.size());
-            if (!s->cdc_options().preimage()) {
-                constexpr int batch_size = 100;
-                int muts_len = muts.size();
-                return transform_mutations(muts, batch_size, [&muts, &trans, batch_size, muts_len] (int idx) {
-                    for (int len = std::min(idx + batch_size, muts_len); idx < len; ++idx) {
-                        muts.push_back(trans.transform(muts[idx]));
-                    }
+} // namespace cdc
+
+future<std::tuple<std::vector<mutation>, cdc::result_callback>>
+cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout, std::vector<mutation>&& mutations) {
+    // we do all this because in the case of batches, we can have mixed schemas.
+    auto e = mutations.end();
+    auto i = std::find_if(mutations.begin(), e, [](const mutation& m) {
+        return m.schema()->cdc_options().enabled();
+    });
+
+    if (i == e) {
+        return make_ready_future<std::tuple<std::vector<mutation>, cdc::result_callback>>(std::make_tuple(std::move(mutations), result_callback{}));
+    }
+
+    mutations.reserve(2 * mutations.size());
+
+    return do_with(std::move(mutations), service::query_state(service::client_state::for_internal_calls(), empty_service_permit()), [this, timeout, i](std::vector<mutation>& mutations, service::query_state& qs) {
+        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs] (int idx) {
+            auto& m = mutations[idx];
+            auto s = m.schema();
+
+            if (!s->cdc_options().enabled()) {
+                return make_ready_future<>();
+            }
+            // for batches/multiple mutations this is super inefficient. either partition the mutation set by schema
+            // and re-use streams, or probably better: add a cache so this lookup is a noop on second mutation
+            return get_streams(_ctxt, s->ks_name(), s->cf_name(), timeout, qs).then([this, s = std::move(s), &qs, &mutations, idx](::shared_ptr<transformer::streams_type> streams) mutable {
+                auto& m = mutations[idx]; // should not really need because of reserve, but lets be conservative
+                transformer trans(_ctxt, s, streams);
+
+                if (!s->cdc_options().preimage()) {
+                    mutations.emplace_back(trans.transform(m));
                     return make_ready_future<>();
+                }
+
+                // Note: further improvement here would be to coalesce the pre-image selects into one
+                // iff a batch contains several modifications to the same table. Otoh, batch is rare(?)
+                // so this is premature.
+                auto f = trans.pre_image_select(qs.get_client_state(), db::consistency_level::LOCAL_QUORUM, m);
+                return f.then([trans = std::move(trans), &mutations, idx] (lw_shared_ptr<cql3::untyped_result_set> rs) mutable {
```

Comment:
How can we be sure that `pre_image_select` doesn't use internals of `trans`? If it so happens that it does, and we yield here, then this `move` into lambda might cause undefined behavior...

Reply:
We know because we wrote it to be. But in general I agree with your sentiment. Of course, I would much rather solve the whole issue by moving to coroutines and stop writing ass-backwards code that moves scope local objects around for every call, but...